src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java

```diff
@@ -181,10 +200,26 @@ private ModifiableRealmIdentity getRealmIdentity(final Principal principal, fina
     }
 
     private DirContext obtainContext() throws RealmUnavailableException {
-        try {
-            return dirContextSupplier.get();
-        } catch (NamingException e) {
-            throw log.ldapRealmFailedToObtainContext(e);
+        if (this.listenersPendingRegistration.size() == 0) {
```

Comment:
Moreover, the big condition is useless here - the `else` part is sufficient (as the condition does not bring any optimization after adding correct synchronization).